optuna/pruners/threshold.py

```diff
@@ -0,0 +1,140 @@
+import math
+from typing import Optional
+
+import optuna
+from optuna.pruners import BasePruner
+from optuna.pruners.percentile import _is_first_in_interval_step
+
+
+class ThresholdPruner(BasePruner):
+    """Pruner to detect outlying metrics of the trials.
+
+    Prune if a metric exceeds upper threshold,
+    falls behind lower threshold or reaches ``nan``.
+
+    Example:
+        .. testcode::
+
+            from optuna import create_study
+            from optuna.exceptions import TrialPruned
+            from optuna.pruners import ThresholdPruner
+
+            def objective_for_upper(trial):
+                for step, y in enumerate(ys_for_upper):
+                    trial.report(y, step)
+
+                    if trial.should_prune():
+                        raise TrialPruned()
+                return ys_for_upper[-1]
+
+
+            def objective_for_lower(trial):
+                for step, y in enumerate(ys_for_lower):
+                    trial.report(y, step)
+
+                    if trial.should_prune():
+                        raise TrialPruned()
+                return ys_for_lower[-1]
+
+
+            ys_for_upper = [0.0, 0.1, 0.2, 0.5, 1.2]
+            ys_for_lower = [100.0, 90.0, 0.1, 0.0, -1]
+            n_trial_step = 5
+
+            study = create_study(pruner=ThresholdPruner(upper=1.0))
+            study.optimize(objective_for_upper, n_trials=10)
+
+            study = create_study(pruner=ThresholdPruner(lower=0.0))
+            study.optimize(objective_for_lower, n_trials=10)
+
+
+    Args
+        lower:
+            A minimum value which determines whether pruner prunes or not.
+            If an intermediate value is smaller than lower, it prunes.
+        upper:
+            A maximum value which determines whether pruner prunes or not.
+            If an intermediate value is larger than upper, it prunes.
+        n_warmup_steps:
+            Pruning is disabled until the trial exceeds the given number of step.
+        interval_steps:
+            Interval in number of steps between the pruning checks, offset by the warmup steps.
+            If no value has been reported at the time of a pruning check, that particular check
+            will be postponed until a value is reported. Value must be at least 1.
+
+    """
+
+    def __init__(
+        self,
+        lower: Optional[float] = None,
+        upper: Optional[float] = None,
+        n_warmup_steps: int = 0,
+        interval_steps: int = 1,
+    ) -> None:
+
+        if lower is not None and not isinstance(lower, float):
+            raise TypeError("lower should be either None or a floating point.")
+        if upper is not None and not isinstance(upper, float):
+            raise TypeError("upper should be either None or a floating point.")
+        if lower is None and upper is None:
+            raise ValueError("Either lower or upper must be specified.")
```

Comment:
The validation looks a lot simpler now. However, I'd like to align the semantics with https://github.com/optuna/optuna/blob/master/optuna/trial.py#L507-L514 in favor of usability as arguments have similar meanings, i.e. try casting to `float` (which would also mean "allowing" `bool`s ). Do you think that's reasonable? If yes, but if it's too much of dragging out, we could address that in a separate PR as it won't break existing code.

Reply:
Thank you, I think it's reasonable.
I revised my PR.

- introduce `_check_value` which is the same as https://github.com/optuna/optuna/blob/master/optuna/trial.py#L507-L514
- use it to check `lower` and `upper`

I'll create another PR to move `_check_value` to a suitable place (if you have a good idea, please let me know) and use it in `trial.py`.
What do you think? @hvy

Reply:
Good question. I'd personally like to create a new lower level submodule (e.g. instead of `study.py` avoiding having several modules depend on `study.py`). It could be private. But I would assume others have other opinions.